cilium/cmd/monitor.go

```diff
@@ -257,35 +256,20 @@ start:
 	}
 
 	var meta payload.Meta
-	metaBuf := make([]byte, unsafe.Sizeof(meta))
+	var pl payload.Payload
 	for {
-		// We need to know about the incoming payload otherwise can't know
-		// how much data to read.
-		if _, err := io.ReadFull(conn, metaBuf); err != nil {
+		if err := payload.ReadMetaPayload(conn, &meta, &pl); err != nil {
 			if err == io.EOF {
+				// EOF may be due to invalid payload size. Close the connection just in case.
+				conn.Close()
+				log.WithError(err).Warn("connection closed")
 				time.Sleep(connTimeout)
 				goto start
+			} else {
+				log.WithError(err).Fatal("decoding error")
```

Comment:
I assume `io.ErrUnexpectedEOF` is intentionally not handled above and treated as a fatal error like all other errors. Thus requiring a manual restart of the monitor. I think that's fine.

Reply:
I didn't think of this error. It seems like I should handle it like EOF above. I'll update the PR.

Reply:
Done.